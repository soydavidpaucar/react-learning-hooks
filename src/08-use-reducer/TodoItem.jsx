export const TodoItem = ({ todo, deleteTodo, toggleTodo }) => {
  return (
    <li className="list-group-item d-flex justify-content-between" onClick={() => toggleTodo(todo.id)}>
      <span className={`align-self-center${todo.done ? ' text-decoration-line-through' : ''}`}>{todo.description}</span>
      <button className="btn btn-danger" onClick={() => deleteTodo(todo.id)}>
        Delete
      </button>
    </li>
  );
};
