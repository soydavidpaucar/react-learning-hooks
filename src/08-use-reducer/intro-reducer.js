const initialState = [
  {
    id: 1,
    todo: 'Collect eggs',
    done: false
  }
];

/**
 * It returns the current state if the action type is not 'add todo', otherwise it returns a new array
 * with the current state and the new todo item
 * @param [state] - The current state of the reducer.
 * @param [action] - This is the action object that was dispatched.
 * @returns The state is being returned.
 */
const todoReducer = (state = initialState, action = {}) => {
  if (action.type == 'add todo') {
    return [...state, action.payload];
  }

  return state;
};

/* Initializing the todos variable with the initial state of the reducer. */
let todos = todoReducer();

/* Creating a new todo object. */
const newTodo = {
  id: 2,
  todo: 'Buy bread',
  done: false
};

/* Creating an action object. */
const addTodoAction = {
  type: 'add todo',
  payload: newTodo
};

/* Dispatching the action object to the reducer. */
todos = todoReducer(todos, addTodoAction);

console.log(todos);
