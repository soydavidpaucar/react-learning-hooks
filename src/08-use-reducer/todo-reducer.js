export const todoReducer = (initialState = [], action) => {
  switch (action.type) {
    case 'add':
      return [...initialState, action.payload];
    case 'delete':
      return initialState.filter((todo) => todo.id !== action.payload);
    case 'toggle':
      return initialState.map((todo) => {
        if (todo.id === action.payload) {
          return {
            ...todo,
            done: !todo.done
          };
        }
        return todo;
      });
    default:
      return initialState;
  }
};
