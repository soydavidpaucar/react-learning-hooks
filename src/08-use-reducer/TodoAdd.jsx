import { useForm } from '../hooks/useForm';

export const TodoAdd = ({ handleNewTodo }) => {
  const { description, handleChanges, handleResetForm } = useForm({
    description: ''
  });

  const handleSubmit = (e) => {
    e.preventDefault();

    if (description.trim().length < 1) {
      return;
    }

    const newTodo = {
      id: new Date().getTime(),
      description,
      done: false
    };

    handleNewTodo(newTodo);
    handleResetForm();
  };

  return (
    <form onSubmit={handleSubmit}>
      <input
        type="text"
        placeholder="Add todo..."
        className="form-control"
        name="description"
        onChange={handleChanges}
        value={description}
      />

      <button type="submit" className="btn btn-outline-primary mt-2 btn-block">
        Submit
      </button>
    </form>
  );
};
