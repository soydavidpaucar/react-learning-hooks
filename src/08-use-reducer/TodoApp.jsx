import { useEffect, useReducer } from 'react';

import { todoReducer } from './todo-reducer';
import { TodoAdd } from './TodoAdd';
import { TodoList } from './TodoList';

const initialState = [
  /* {
    id: new Date().getTime(),
    description: 'Learn React',
    done: false
  } */
];

/**
 * If there are todos in localStorage, parse them and return them, otherwise return an empty array
 * @returns An array of todos.
 */
const init = () => {
  return JSON.parse(localStorage.getItem('todos')) || [];
};

export const TodoApp = () => {
  const [todos, dispatch] = useReducer(todoReducer, initialState, init);

  /* Saving the todos in localStorage. */
  useEffect(() => {
    localStorage.setItem('todos', JSON.stringify(todos));
  }, [todos]);

  const handleNewTodo = (newTodo) => {
    const action = {
      type: 'add',
      payload: newTodo
    };

    dispatch(action);
  };

  return (
    <>
      <h1>
        TodoApp {todos.length}, <small>Pending: 2</small>
      </h1>
      <hr />

      <div className="row">
        <div className="col-7">
          <TodoList todos={todos} />
        </div>
        <div className="col-5">
          <h4>Add Todo</h4>
          <hr />
          <TodoAdd handleNewTodo={handleNewTodo} />
        </div>
      </div>
    </>
  );
};
