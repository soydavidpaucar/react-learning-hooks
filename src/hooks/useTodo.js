import { useEffect, useReducer } from 'react';

import { todoReducer } from '../08-use-reducer/todo-reducer';

/**
 * If there's a todos array in localStorage, return it, otherwise return an empty array
 * @returns An array of todos.
 */
const init = () => {
  return JSON.parse(localStorage.getItem('todos')) || [];
};

export const useTodo = () => {
  const [todos, dispatch] = useReducer(todoReducer, [], init);

  const totalTodos = todos.length;
  const pendingTodos = todos.filter((todo) => !todo.done).length;

  useEffect(() => {
    localStorage.setItem('todos', JSON.stringify(todos));
  }, [todos]);

  const handleNewTodo = (newTodo) => {
    const action = {
      type: 'add',
      payload: newTodo
    };

    dispatch(action);
  };

  const handleDeleteTodo = (todoId) => {
    const action = {
      type: 'delete',
      payload: todoId
    };

    dispatch(action);
  };

  const handleToggleTodo = (todoId) => {
    const action = {
      type: 'toggle',
      payload: todoId
    };

    dispatch(action);
  };

  return {
    todos,
    totalTodos,
    pendingTodos,
    handleNewTodo,
    handleDeleteTodo,
    handleToggleTodo
  };
};
