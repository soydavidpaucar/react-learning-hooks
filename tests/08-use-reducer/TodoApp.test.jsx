import { render, screen } from '@testing-library/react';
import { describe, expect, it, vi } from 'vitest';

import { TodoApp } from '../../src/08-use-reducer/TodoApp';
import { useTodo } from '../../src/hooks/useTodo.js';

vi.mock('../../src/hooks/useTodo.js');

describe('test on TodoApp', () => {
  it('should match snapshot', () => {
    vi.mocked(useTodo).mockReturnValue({
      todos: [
        {
          id: new Date().getTime(),
          description: 'Learn React',
          done: false
        },
        {
          id: new Date().getTime() + 1,
          description: 'Learn Vue',
          done: true
        }
      ],
      totalTodos: 2,
      pendingTodos: 1,
      handleNewTodo: vi.fn(),
      handleDeleteTodo: vi.fn(),
      handleToggleTodo: vi.fn()
    });

    const { container } = render(<TodoApp />);

    expect(container).toMatchSnapshot();

    expect(useTodo).toHaveBeenCalledTimes(1);
    expect(screen.getByText('Learn React')).toBeTruthy();
    expect(screen.getByText('Learn Vue')).toBeTruthy();
  });
});
