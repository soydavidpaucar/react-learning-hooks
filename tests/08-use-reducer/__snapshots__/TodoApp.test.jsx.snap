// Vitest Snapshot v1

exports[`test on TodoApp > should match snapshot 1`] = `
<div>
  <h1>
    TodoApp 
    2
    , 
    <small>
      Pending: 
      1
    </small>
  </h1>
  <hr />
  <div
    class="row"
  >
    <div
      class="col-7"
    >
      <ul
        class="list-group"
      >
        <li
          class="list-group-item d-flex justify-content-between"
        >
          <span
            class="align-self-center"
          >
            Learn React
          </span>
          <button
            class="btn btn-danger"
          >
            Delete
          </button>
        </li>
        <li
          class="list-group-item d-flex justify-content-between"
        >
          <span
            class="align-self-center text-decoration-line-through"
          >
            Learn Vue
          </span>
          <button
            class="btn btn-danger"
          >
            Delete
          </button>
        </li>
      </ul>
    </div>
    <div
      class="col-5"
    >
      <h4>
        Add Todo
      </h4>
      <hr />
      <form>
        <input
          class="form-control"
          name="description"
          placeholder="Add todo..."
          type="text"
          value=""
        />
        <button
          class="btn btn-outline-primary mt-2 btn-block"
          type="submit"
        >
          Submit
        </button>
      </form>
    </div>
  </div>
</div>
`;
